File: final_assignment_backend_quarkus/src/main/java/finalassignmentbackend/config/docker/RunDocker.java
package finalassignmentbackend.config.docker;

import io.quarkus.runtime.ShutdownEvent;
import io.quarkus.runtime.StartupEvent;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.enterprise.event.Observes;
import jakarta.inject.Inject;
import org.eclipse.microprofile.config.inject.ConfigProperty;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.testcontainers.containers.GenericContainer;
import org.testcontainers.containers.wait.strategy.Wait;
import org.testcontainers.utility.DockerImageName;

import java.time.Duration;

@ApplicationScoped
public class RunDocker {

    private static final Logger log = LoggerFactory.getLogger(RunDocker.class);

    private GenericContainer<?> elasticsearchContainer;
    private GenericContainer<?> redisContainer;
    private GenericContainer<?> redpandaContainer;

    @Inject
    @ConfigProperty(name = "elasticsearch.image", defaultValue = "docker.elastic.co/elasticsearch/elasticsearch:9.4.1")
    String elasticsearchImage;

    @Inject
    @ConfigProperty(name = "redis.image", defaultValue = "redis:7")
    String redisImage;

    @Inject
    @ConfigProperty(name = "redpanda.image", defaultValue = "docker.redpanda.com/redpandadata/redpanda:v26.1.9")
    String redpandaImage;

    void onStartElasticsearch(@Observes StartupEvent event) {
        startElasticsearch();
    }

    void onStopElasticsearch(@Observes ShutdownEvent event) {
        stopElasticsearch();
    }

    void onStartRedis(@Observes StartupEvent event) {
        startRedis();
    }

    void onStopRedis(@Observes ShutdownEvent event) {
        stopRedis();
    }

    void onStartRedpanda(@Observes StartupEvent event) {
        startRedpanda();
    }

    void onStopRedpanda(@Observes ShutdownEvent event) {
        stopRedpanda();
    }

    @SuppressWarnings("resource")
    public void startElasticsearch() {
        try {
            elasticsearchContainer = new GenericContainer<>(DockerImageName.parse(elasticsearchImage))
                    .withExposedPorts(9200)
                    .withEnv("xpack.security.enabled", "false")
                    .withEnv("discovery.type", "single-node")
                    .waitingFor(Wait.forHttp("/_cluster/health")
                            .forPort(9200)
                            .withStartupTimeout(Duration.ofSeconds(120)));
            elasticsearchContainer.start();

            String host = elasticsearchContainer.getHost();
            Integer httpPort = elasticsearchContainer.getMappedPort(9200);
            String elasticsearchUrl = String.format("http://%s:%d", host, httpPort);
            System.setProperty("elasticsearch.host", elasticsearchUrl);
            log.info("Elasticsearch container started successfully at {}", elasticsearchUrl);
        } catch (Exception e) {
            log.warn("Failed to start Elasticsearch container: {}", e.getMessage());
        }
    }

    @SuppressWarnings("resource")
    public void startRedis() {
        try {
            redisContainer = new GenericContainer<>(DockerImageName.parse(redisImage))
                    .withExposedPorts(6379)
                    .waitingFor(Wait.forListeningPort().withStartupTimeout(Duration.ofSeconds(60)));
            redisContainer.start();

            String redisHost = redisContainer.getHost();
            int redisPort = redisContainer.getMappedPort(6379);
            System.setProperty("quarkus.redis.hosts", redisHost + ":" + redisPort);
            log.info("Redis container started successfully at {}:{}", redisHost, redisPort);
        } catch (Exception e) {
            log.warn("Failed to start Redis container: {}", e.getMessage());
        }
    }

    @SuppressWarnings("resource")
    public void startRedpanda() {
        try {
            redpandaContainer = new GenericContainer<>(DockerImageName.parse(redpandaImage))
                    .withExposedPorts(9092)
                    .waitingFor(Wait.forListeningPort().withStartupTimeout(Duration.ofSeconds(120)));
            redpandaContainer.start();

            String host = redpandaContainer.getHost();
            int port = redpandaContainer.getMappedPort(9092);
            System.setProperty("quarkus.kafka.bootstrap.servers", host + ":" + port);
            log.info("Redpanda container started successfully at {}:{}", host, port);
        } catch (Exception e) {
            log.warn("Failed to start Redpanda container: {}", e.getMessage());
        }
    }

    public void stopElasticsearch() {
        stopContainer(elasticsearchContainer, "Elasticsearch");
    }

    public void stopRedis() {
        stopContainer(redisContainer, "Redis");
    }

    public void stopRedpanda() {
        stopContainer(redpandaContainer, "Redpanda");
    }

    private void stopContainer(GenericContainer<?> container, String name) {
        if (container != null && container.isRunning()) {
            container.stop();
            container.close();
            log.info("{} container stopped and closed", name);
        }
    }
}

File: final_assignment_backend_go/project/configs/docker/run_docker.go
package docker

import (
	"context"
	"fmt"
	"log"
	"os"
	"time"

	"github.com/testcontainers/testcontainers-go"
	"github.com/testcontainers/testcontainers-go/modules/elasticsearch"
	"github.com/testcontainers/testcontainers-go/modules/redis"
	"github.com/testcontainers/testcontainers-go/wait"
)

// RunDocker 缁撴瀯浣撶敤浜庣鐞嗘墍鏈夊鍣?
type RunDocker struct {
	ctx                    context.Context
	redisContainer         *redis.RedisContainer
	redpandaContainer      testcontainers.Container
	elasticsearchContainer *elasticsearch.ElasticsearchContainer
}

// NewRunDocker 鍒濆鍖栦竴涓柊鐨勫疄渚?
func NewRunDocker() *RunDocker {
	return &RunDocker{ctx: context.Background()}
}

// Init 鍚姩鎵€鏈夊鍣?
func (r *RunDocker) Init() {
	r.startRedis()
	r.startRedpanda()
	r.startElasticsearch()
}

// startRedis 鍚姩 Redis 瀹瑰櫒
func (r *RunDocker) startRedis() {
	log.Println("[INFO] Starting Redis container...")
	container, err := redis.RunContainer(r.ctx,
		testcontainers.WithImage("redis:7"),
	)
	if err != nil {
		log.Printf("[ERROR] Failed to start Redis: %v\n", err)
		return
	}

	host, _ := container.Host(r.ctx)
	port, _ := container.MappedPort(r.ctx, "6379/tcp")

	err = os.Setenv("REDIS_HOST", host)
	if err != nil {
		return
	}

	err = os.Setenv("REDIS_PORT", port.Port())
	if err != nil {
		return
	}

	log.Printf("[INFO] Redis started at %s:%s\n", host, port.Port())

	r.redisContainer = container
}

// startRedpanda 鍚姩 Redpanda 瀹瑰櫒
func (r *RunDocker) startRedpanda() {
	log.Println("[INFO] Starting Redpanda container...")
	req := testcontainers.ContainerRequest{
		Image:        "redpandadata/redpanda:v24.1.2",
		ExposedPorts: []string{"9092/tcp"},
		WaitingFor:   wait.ForLog("Started Kafka API server").WithStartupTimeout(2 * time.Minute),
	}
	container, err := testcontainers.GenericContainer(r.ctx, testcontainers.GenericContainerRequest{
		ContainerRequest: req,
		Started:          true,
	})
	if err != nil {
		log.Printf("[ERROR] Failed to start Redpanda: %v\n", err)
		return
	}

	host, _ := container.Host(r.ctx)
	port, _ := container.MappedPort(r.ctx, "9092/tcp")
	bootstrap := fmt.Sprintf("%s:%s", host, port.Port())

	err = os.Setenv("KAFKA_BOOTSTRAP_SERVERS", bootstrap)
	if err != nil {
		return
	}

	log.Printf("[INFO] Redpanda started, bootstrap servers: %s\n", bootstrap)

	r.redpandaContainer = container
}

// startElasticsearch 鍚姩 Elasticsearch 瀹瑰櫒
func (r *RunDocker) startElasticsearch() {
	log.Println("[INFO] Starting Elasticsearch container...")

	container, err := elasticsearch.RunContainer(r.ctx,
		testcontainers.WithImage("docker.elastic.co/elasticsearch/elasticsearch:9.4.1"),
		testcontainers.WithEnv(map[string]string{
			"xpack.security.enabled": "false",
			"discovery.type":         "single-node",
		}),
	)
	if err != nil {
		log.Printf("[ERROR] Failed to start Elasticsearch: %v\n", err)
		return
	}

	host, _ := container.Host(r.ctx)
	port, _ := container.MappedPort(r.ctx, "9200/tcp")

	url := fmt.Sprintf("http://%s:%s", host, port.Port())

	err = os.Setenv("ELASTICSEARCH_URL", url)
	if err != nil {
		return
	}

	log.Printf("[INFO] Elasticsearch started at: %s\n", url)
	r.elasticsearchContainer = container
}

// StopContainers 鍋滄鎵€鏈夊鍣?
func (r *RunDocker) StopContainers() {
	if r.redisContainer != nil {
		err := r.redisContainer.Terminate(r.ctx)
		if err != nil {
			return
		}
		log.Println("[INFO] Redis container stopped")
	}
	if r.redpandaContainer != nil {
		err := r.redpandaContainer.Terminate(r.ctx)
		if err != nil {
			return
		}
		log.Println("[INFO] Redpanda container stopped")
	}
	if r.elasticsearchContainer != nil {

		err := r.elasticsearchContainer.Terminate(r.ctx)
		if err != nil {
			return
		}
		log.Println("[INFO] Elasticsearch container stopped")
	}
}
